File: pdk/python/prx_pdk/host.py
"""
prx_pdk.host — Host function wrappers for PRX WASM plugins.

When running inside a WASM component (built with componentize-py), these
functions call the real host via WIT-generated bindings.

When imported in a regular Python environment (local development / testing),
they fall back to safe stubs so that plugin code can be unit-tested without
a WASM runtime.

Usage::

    from prx_pdk import host

    host.log.info("Plugin started")
    value = host.config.get("timeout_ms")
    host.kv.set("last_run", b"2025-01-01")
"""

from __future__ import annotations

import json
import os
import sys
import time
from typing import Optional

from .types import HttpResponse, MemoryEntry

# ── Detect WASM environment ───────────────────────────────────────────────────
# componentize-py sets sys.platform == "wasi" and generates the
# `prx.host.*` package as WIT bindings.

_WASM = sys.platform == "wasi"

if _WASM:
    try:
        from prx.host import log as _wit_log          # type: ignore[import]
        from prx.host import config as _wit_config    # type: ignore[import]
        from prx.host import kv as _wit_kv            # type: ignore[import]
        from prx.host import events as _wit_events    # type: ignore[import]
        from prx.host import http_outbound as _wit_http  # type: ignore[import]
        from prx.host import memory as _wit_memory    # type: ignore[import]
        _BINDINGS_AVAILABLE = True
    except ImportError:
        _BINDINGS_AVAILABLE = False
else:
    _BINDINGS_AVAILABLE = False


# ── log ───────────────────────────────────────────────────────────────────────

class _Log:
    """Structured logging — writes to the PRX tracing infrastructure.

    Log messages appear in the host's log output with the plugin name as
    context.

    WIT interface: ``prx:host/log``
    """

    def _emit(self, level: str, msg: str) -> None:
        if _BINDINGS_AVAILABLE:
            _wit_log.log(level, msg)
        else:
            tag = level.upper().ljust(5)
            print(f"[prx-pdk {tag}] {msg}", file=sys.stderr)

    def info(self, msg: str) -> None:
        """Emit an INFO-level log message."""
        self._emit("info", msg)

    def warn(self, msg: str) -> None:
        """Emit a WARN-level log message."""
        self._emit("warn", msg)

    def error(self, msg: str) -> None:
        """Emit an ERROR-level log message."""
        self._emit("error", msg)

    def debug(self, msg: str) -> None:
        """Emit a DEBUG-level log message."""
        self._emit("debug", msg)

    def trace(self, msg: str) -> None:
        """Emit a TRACE-level log message."""
        self._emit("trace", msg)


log = _Log()


# ── config ────────────────────────────────────────────────────────────────────

class _Config:
    """Plugin configuration — read-only access to values from ``plugin.toml [config]``.

    Config values are set by the operator when deploying the plugin and cannot
    be modified at runtime.  Use :mod:`kv` for mutable persistent storage.

    WIT interface: ``prx:host/config``
    """

    def get(self, key: str) -> Optional[str]:
        """Get a configuration value by key. Returns ``None`` if not set."""
        if _BINDINGS_AVAILABLE:
            return _wit_config.get(key)
        # Local stub: fall back to environment variables
        return os.environ.get(key)

    def get_all(self) -> list[tuple[str, str]]:
        """Get all configuration key-value pairs."""
        if _BINDINGS_AVAILABLE:
            return list(_wit_config.get_all())
        return []

    def get_or(self, key: str, default: str) -> str:
        """Get a configuration value, returning *default* if not set."""
        v = self.get(key)
        return v if v is not None else default


config = _Config()


# ── kv ───────────────────────────────────────────────────────────────────────

class _Kv:
    """Key-value storage — isolated per-plugin persistent store.

    Each plugin gets its own namespace; plugins cannot access each other's keys.
    Values are raw bytes.  Use :meth:`get_json` / :meth:`set_json` for
    structured data.

    WIT interface: ``prx:host/kv``
    """

    # Local-dev in-memory store (only used when not in WASM)
    _store: dict[str, bytes] = {}

    def get(self, key: str) -> Optional[bytes]:
        """Retrieve a value by key. Returns ``None`` if the key does not exist."""
        if _BINDINGS_AVAILABLE:
            return _wit_kv.get(key)
        return self._store.get(key)

    def get_str(self, key: str) -> Optional[str]:
        """Retrieve and decode as UTF-8 text."""
        v = self.get(key)
        if v is None:
            return None
        return v.decode("utf-8")

    def get_json(self, key: str) -> object:
        """Retrieve and JSON-deserialise a stored value. Returns ``None`` if missing."""
        v = self.get(key)
        if v is None:
            return None
        return json.loads(v)

    def set(self, key: str, value: bytes) -> None:
        """Store a byte value. Overwrites any existing value."""
        if _BINDINGS_AVAILABLE:
            _wit_kv.set(key, value)
        else:
            self._store[key] = value

    def set_str(self, key: str, value: str) -> None:
        """Store a UTF-8 string value."""
        self.set(key, value.encode("utf-8"))

    def set_json(self, key: str, value: object) -> None:
        """JSON-serialise and store a value."""
        self.set(key, json.dumps(value).encode("utf-8"))

    def delete(self, key: str) -> bool:
        """Delete a key. Returns ``True`` if the key existed."""
        if _BINDINGS_AVAILABLE:
            return _wit_kv.delete(key)
        existed = key in self._store
        self._store.pop(key, None)
        return existed

    def list_keys(self, prefix: str) -> list[str]:
        """List all keys matching a prefix."""
        if _BINDINGS_AVAILABLE:
            return list(_wit_kv.list_keys(prefix))
        return [k for k in self._store if k.startswith(prefix)]

    def increment(self, key: str, delta: int = 1) -> int:
        """Atomically increment an integer counter stored at *key*.

        Initialises to 0 if the key does not exist, then adds *delta*.
        """
        current_bytes = self.get(key)
        if current_bytes is None:
            current = 0
        else:
            try:
                current = json.loads(current_bytes)
                if not isinstance(current, int):
                    raise ValueError(f"not an int: {current!r}")
            except (ValueError, json.JSONDecodeError) as exc:
                log.warn(f"kv.increment: key '{key}' exists but is not a valid int "
                         f"(resetting to 0): {exc}")
                current = 0
        next_val = current + delta
        self.set_json(key, next_val)
        return next_val


kv = _Kv()


# ── events ────────────────────────────────────────────────────────────────────

class _Events:
    """Event bus — fire-and-forget publish/subscribe for inter-plugin communication.

    Events flow through the host for auditing and access control.
    Payload must be valid JSON, max 64 KB.

    WIT interface: ``prx:host/events``
    """

    def publish(self, topic: str, payload: str) -> None:
        """Publish an event to a topic.

        All subscribers matching the topic will receive the event asynchronously.

        :raises RuntimeError: if the plugin lacks ``events`` permission or the
            payload exceeds 64 KB.
        """
        if _BINDINGS_AVAILABLE:
            result = _wit_events.publish(topic, payload)
            if isinstance(result, Exception):
                raise RuntimeError(str(result))
        else:
            log.debug(f"[stub] events.publish topic={topic!r} payload={payload!r}")

    def publish_json(self, topic: str, payload: object) -> None:
        """Publish a JSON-serialisable value to a topic."""
        self.publish(topic, json.dumps(payload))

    def subscribe(self, pattern: str) -> int:
        """Subscribe to a topic pattern.

        Supports exact match (``"weather.update"``) and wildcard
        (``"weather.*"``).  Returns a subscription ID for use with
        :meth:`unsubscribe`.
        """
        if _BINDINGS_AVAILABLE:
            return _wit_events.subscribe(pattern)
        log.debug(f"[stub] events.subscribe pattern={pattern!r}")
        return 0

    def unsubscribe(self, subscription_id: int) -> None:
        """Cancel a subscription by ID."""
        if _BINDINGS_AVAILABLE:
            _wit_events.unsubscribe(subscription_id)
        else:
            log.debug(f"[stub] events.unsubscribe id={subscription_id}")


events = _Events()


# ── http ──────────────────────────────────────────────────────────────────────

class _Http:
    """Outbound HTTP — make controlled HTTP requests from plugins.

    URLs are validated against the plugin's ``http_allowlist`` in ``plugin.toml``.
    Requires ``"http-outbound"`` permission.

    WIT interface: ``prx:host/http-outbound``
    """

    def request(
        self,
        method: str,
        url: str,
        headers: list[tuple[str, str]] | None = None,
        body: bytes | None = None,
    ) -> HttpResponse:
        """Make an HTTP request.

        :param method: HTTP verb (``"GET"``, ``"POST"``, etc.)
        :param url: Target URL (must be in the plugin's ``http_allowlist``)
        :param headers: Request headers as ``(name, value)`` pairs
        :param body: Optional request body bytes
        :raises RuntimeError: if the URL is not allowed, or the plugin lacks
            ``"http-outbound"`` permission.
        """
        if headers is None:
            headers = []
        if _BINDINGS_AVAILABLE:
            result = _wit_http.request(method, url, headers, body)
            return HttpResponse(
                status=result.status,
                headers=list(result.headers),
                body=bytes(result.body),
            )
        raise RuntimeError(
            "http.request is only available inside a WASM component. "
            "Use a mock in your tests."
        )

    def get(self, url: str, headers: list[tuple[str, str]] | None = None) -> HttpResponse:
        """Convenience: HTTP GET request."""
        return self.request("GET", url, headers or [])

    def post_json(
        self,
        url: str,
        payload: object,
        headers: list[tuple[str, str]] | None = None,
    ) -> HttpResponse:
        """Convenience: HTTP POST with a JSON body."""
        h = list(headers or [])
        if not any(k.lower() == "content-type" for k, _ in h):
            h.append(("Content-Type", "application/json"))
        return self.request("POST", url, h, json.dumps(payload).encode())


http = _Http()


# ── clock ─────────────────────────────────────────────────────────────────────

class _Clock:
    """Clock — current time utilities for plugins.

    Note: The PRX WIT spec does not currently expose a dedicated clock interface.
    In WASM, WASI time functions are used.  On the host, ``time.time()`` is used.
    """

    def now_ms(self) -> int:
        """Return the current time as Unix milliseconds (UTC)."""
        return int(time.time() * 1000)

    def timezone(self) -> str:
        """Return the host timezone name.

        Currently always returns ``"UTC"`` — timezone support is planned for a
        future PRX host interface release.
        """
        return "UTC"


clock = _Clock()


# ── memory ────────────────────────────────────────────────────────────────────

class _Memory:
    """Long-term memory — store and recall text entries.

    Requires ``"memory"`` permission in ``plugin.toml``.

    WIT interface: ``prx:host/memory``
    """

    def store(self, text: str, category: str) -> str:
        """Store text in memory. Returns the generated entry ID."""
        if _BINDINGS_AVAILABLE:
            return _wit_memory.store(text, category)
        log.debug(f"[stub] memory.store category={category!r} text={text!r}")
        return "stub-id"

    def recall(self, query: str, limit: int = 10) -> list[MemoryEntry]:
        """Recall memories matching a query. Returns up to *limit* entries."""
        if _BINDINGS_AVAILABLE:
            raw = _wit_memory.recall(query, limit)
            return [
                MemoryEntry(
                    id=e.id,
                    text=e.text,
                    category=e.category,
                    importance=e.importance,
                )
                for e in raw
            ]
        return []


memory = _Memory()

File: pdk/javascript/packages/prx-pdk/src/host.ts
/**
 * PRX PDK — Host function wrappers.
 *
 * This module provides ergonomic TypeScript wrappers around the WIT-generated
 * host function bindings exposed by the PRX runtime.
 *
 * ## WASM vs non-WASM environments
 *
 * When compiled to a WASM component via `jco componentize`, the PRX runtime
 * injects the real host implementations as WIT imports.  In that environment
 * the wrapper functions delegate to the generated bindings.
 *
 * Outside a WASM component (e.g. during unit tests or local development),
 * the WIT bindings are unavailable.  The wrappers fall back to harmless stubs
 * (console.log for logging, no-ops for storage/events) so that plugin code can
 * be tested without a running PRX host.
 *
 * Detection is performed via `globalThis.__prx_host__`, which the
 * jco-componentize runtime sets to the bound host exports object.
 */

import type { HttpResponse, MemoryEntry, LogLevel } from "./types.js";

// ── Internal: access WIT host bindings ───────────────────────────────────────

/**
 * Shape of the WIT host bindings object injected by jco componentize.
 *
 * This mirrors the namespace structure produced by `jco transpile` for the
 * `prx:host` package:
 *   - `prx:host/log`
 *   - `prx:host/config`
 *   - `prx:host/kv`
 *   - `prx:host/events`
 *   - `prx:host/http-outbound`
 *   - `prx:host/memory`
 */
interface PrxHostBindings {
  "prx:host/log": {
    log(level: string, message: string): void;
  };
  "prx:host/config": {
    get(key: string): string | undefined;
    getAll(): [string, string][];
  };
  "prx:host/kv": {
    get(key: string): Uint8Array | undefined;
    set(key: string, value: Uint8Array): void;
    delete(key: string): boolean;
    listKeys(prefix: string): string[];
  };
  "prx:host/events": {
    publish(topic: string, payload: string): void;
    subscribe(topicPattern: string): bigint;
    unsubscribe(subscriptionId: bigint): void;
  };
  "prx:host/http-outbound": {
    request(
      method: string,
      url: string,
      headers: [string, string][],
      body: Uint8Array | undefined,
    ): { status: number; headers: [string, string][]; body: Uint8Array };
  };
  "prx:host/memory": {
    store(text: string, category: string): string;
    recall(
      query: string,
      limit: number,
    ): { id: string; text: string; category: string; importance: number }[];
  };
}

/** Whether we are running inside a WIT-componentized WASM module. */
function isWasmEnv(): boolean {
  return (
    typeof globalThis !== "undefined" &&
    "__prx_host__" in (globalThis as object) &&
    (globalThis as unknown as Record<string, unknown>)["__prx_host__"] !== null
  );
}

/** Retrieve the injected host bindings (WASM env only). */
function hostBindings(): PrxHostBindings {
  // Cast through unknown to avoid strict globalThis index-signature errors.
  // The __prx_host__ property is only present when running inside a
  // jco-componentized WASM module, where the PRX runtime injects it.
  return (globalThis as unknown as { __prx_host__: PrxHostBindings }).__prx_host__;
}

// ── log ───────────────────────────────────────────────────────────────────────

/**
 * Structured logging — writes to the PRX tracing infrastructure.
 *
 * Outside a WASM environment, log calls are forwarded to `console` so that
 * plugin code is testable locally.
 */
export const log = {
  /** Emit a TRACE-level log message. */
  trace(message: string): void {
    _log("trace", message);
  },
  /** Emit a DEBUG-level log message. */
  debug(message: string): void {
    _log("debug", message);
  },
  /** Emit an INFO-level log message. */
  info(message: string): void {
    _log("info", message);
  },
  /** Emit a WARN-level log message. */
  warn(message: string): void {
    _log("warn", message);
  },
  /** Emit an ERROR-level log message. */
  error(message: string): void {
    _log("error", message);
  },
} as const;

function _log(level: LogLevel, message: string): void {
  if (isWasmEnv()) {
    hostBindings()["prx:host/log"].log(level, message);
  } else {
    // Stub: map to the appropriate console method
    const methods: Record<LogLevel, (msg: string) => void> = {
      trace: (m) => console.debug(`[prx-pdk TRACE] ${m}`),
      debug: (m) => console.debug(`[prx-pdk DEBUG] ${m}`),
      info: (m) => console.info(`[prx-pdk INFO ] ${m}`),
      warn: (m) => console.warn(`[prx-pdk WARN ] ${m}`),
      error: (m) => console.error(`[prx-pdk ERROR] ${m}`),
    };
    methods[level](message);
  }
}

// ── config ────────────────────────────────────────────────────────────────────

/**
 * Plugin configuration — read-only access to values from `plugin.toml [config]`.
 *
 * Config values are set by the operator at deploy time.  Use `kv` for mutable
 * persistent storage.
 */
export const config = {
  /**
   * Get a configuration value by key.
   * Returns `undefined` if the key is not set.
   */
  get(key: string): string | undefined {
    if (isWasmEnv()) {
      return hostBindings()["prx:host/config"].get(key);
    }
    return undefined;
  },

  /** Get all configuration key-value pairs. */
  getAll(): [string, string][] {
    if (isWasmEnv()) {
      return hostBindings()["prx:host/config"].getAll();
    }
    return [];
  },

  /**
   * Get a configuration value, returning `defaultValue` if not set.
   */
  getOr(key: string, defaultValue: string): string {
    return config.get(key) ?? defaultValue;
  },
} as const;

// ── kv ────────────────────────────────────────────────────────────────────────

/**
 * Key-value storage — isolated per-plugin persistent store.
 *
 * Each plugin gets its own namespace; plugins cannot access each other's keys.
 * Values are raw bytes (`Uint8Array`).  Use JSON serialisation for structured data.
 */
export const kv = {
  /**
   * Retrieve a value by key.
   * Returns `undefined` if the key does not exist.
   */
  get(key: string): Uint8Array | undefined {
    if (isWasmEnv()) {
      return hostBindings()["prx:host/kv"].get(key);
    }
    return undefined;
  },

  /** Retrieve a value and decode it as UTF-8 text. */
  getString(key: string): string | undefined {
    const bytes = kv.get(key);
    if (bytes === undefined) return undefined;
    return new TextDecoder().decode(bytes);
  },

  /** Retrieve and JSON-parse a stored value. */
  getJson<T>(key: string): T | undefined {
    const str = kv.getString(key);
    if (str === undefined) return undefined;
    try {
      return JSON.parse(str) as T;
    } catch {
      return undefined;
    }
  },

  /** Store a byte value. Overwrites any existing value for the key. */
  set(key: string, value: Uint8Array): void {
    if (isWasmEnv()) {
      hostBindings()["prx:host/kv"].set(key, value);
    }
    // stub: no-op outside WASM
  },

  /** Store a UTF-8 string value. */
  setString(key: string, value: string): void {
    kv.set(key, new TextEncoder().encode(value));
  },

  /** JSON-serialise and store a value. */
  setJson(key: string, value: unknown): void {
    kv.setString(key, JSON.stringify(value));
  },

  /**
   * Delete a key.
   * Returns `true` if the key existed, `false` otherwise.
   */
  delete(key: string): boolean {
    if (isWasmEnv()) {
      return hostBindings()["prx:host/kv"].delete(key);
    }
    return false;
  },

  /** List all keys matching a prefix. */
  listKeys(prefix: string): string[] {
    if (isWasmEnv()) {
      return hostBindings()["prx:host/kv"].listKeys(prefix);
    }
    return [];
  },

  /**
   * Atomically increment an integer counter stored at `key`.
   *
   * Initialises to 0 if the key does not exist, then adds `delta`.
   * Returns the new value.
   */
  increment(key: string, delta: number): number {
    const current = kv.getJson<number>(key) ?? 0;
    const next =
      typeof current === "number" && isFinite(current) ? current + delta : delta;
    kv.setJson(key, next);
    return next;
  },
} as const;

// ── events ────────────────────────────────────────────────────────────────────

/**
 * Event bus — fire-and-forget publish/subscribe for inter-plugin communication.
 *
 * Events flow through the host for auditing and access control.
 * Requires `"events"` permission in `plugin.toml`.
 * Payload must be valid JSON, max 64 KB.
 */
export const events = {
  /**
   * Publish an event to a topic.
   *
   * All subscribers matching the topic receive the event asynchronously.
   *
   * @param topic  Event topic (e.g. `"weather.update"`).
   * @param payload  JSON-encoded payload string (max 64 KB).
   */
  publish(topic: string, payload: string): void {
    if (isWasmEnv()) {
      hostBindings()["prx:host/events"].publish(topic, payload);
    }
    // stub: log to console so tests can observe published events
    else {
      console.debug(`[prx-pdk events.publish] topic=${topic} payload=${payload}`);
    }
  },

  /**
   * Publish a JSON-serialisable value to a topic.
   */
  publishJson(topic: string, payload: unknown): void {
    events.publish(topic, JSON.stringify(payload));
  },

  /**
   * Subscribe to a topic pattern.
   *
   * Supports exact match (`"weather.update"`) and wildcard (`"weather.*"`).
   * Returns a subscription ID for later `unsubscribe()`.
   */
  subscribe(topicPattern: string): bigint {
    if (isWasmEnv()) {
      return hostBindings()["prx:host/events"].subscribe(topicPattern);
    }
    return 0n;
  },

  /**
   * Cancel a subscription by ID.
   */
  unsubscribe(subscriptionId: bigint): void {
    if (isWasmEnv()) {
      hostBindings()["prx:host/events"].unsubscribe(subscriptionId);
    }
  },
} as const;

// ── http ──────────────────────────────────────────────────────────────────────

/**
 * Outbound HTTP — make controlled HTTP requests from plugins.
 *
 * URLs are validated against the plugin's `http_allowlist` in `plugin.toml`.
 * Requires `"http-outbound"` permission.
 */
export const http = {
  /**
   * Make an HTTP request.
   *
   * @param method   HTTP verb (`"GET"`, `"POST"`, etc.)
   * @param url      Target URL (must match the plugin's `http_allowlist`)
   * @param headers  Request headers as `[name, value]` pairs
   * @param body     Optional request body bytes
   */
  request(
    method: string,
    url: string,
    headers: [string, string][],
    body?: Uint8Array,
  ): HttpResponse {
    if (isWasmEnv()) {
      const raw = hostBindings()["prx:host/http-outbound"].request(
        method,
        url,
        headers,
        body,
      );
      return {
        status: raw.status,
        headers: raw.headers,
        body: raw.body,
      };
    }
    throw new Error(
      "http.request is only available inside a PRX WASM component",
    );
  },

  /** Convenience wrapper: HTTP GET. */
  get(url: string, headers: [string, string][] = []): HttpResponse {
    return http.request("GET", url, headers);
  },

  /** Convenience wrapper: HTTP POST with a JSON body. */
  postJson(
    url: string,
    payload: unknown,
    headers: [string, string][] = [],
  ): HttpResponse {
    const body = new TextEncoder().encode(JSON.stringify(payload));
    const mergedHeaders: [string, string][] = [
      ...headers.filter(([k]) => k.toLowerCase() !== "content-type"),
      ["Content-Type", "application/json"],
    ];
    return http.request("POST", url, mergedHeaders, body);
  },

  /**
   * Parse an `HttpResponse` body as UTF-8 text.
   * Throws if the body is not valid UTF-8.
   */
  bodyText(response: HttpResponse): string {
    return new TextDecoder().decode(response.body);
  },

  /**
   * Parse an `HttpResponse` body as JSON.
   * Throws if the body is not valid UTF-8 or not valid JSON.
   */
  bodyJson<T>(response: HttpResponse): T {
    return JSON.parse(http.bodyText(response)) as T;
  },
} as const;

// ── clock ─────────────────────────────────────────────────────────────────────

/**
 * Clock — current time utilities for plugins.
 *
 * The PRX WIT spec does not expose a dedicated clock interface.  This module
 * provides a best-effort implementation using `Date.now()` which is available
 * in both the jco-componentize WASM sandbox and standard Node.js / browser
 * environments.
 */
export const clock = {
  /** Return the current time as Unix milliseconds (UTC). */
  nowMs(): number {
    return Date.now();
  },

  /** Return the current time as an ISO 8601 string (UTC). */
  nowIso(): string {
    return new Date().toISOString();
  },
} as const;

// ── memory ────────────────────────────────────────────────────────────────────

/**
 * Long-term memory — store and recall text entries.
 *
 * Requires `"memory"` permission in `plugin.toml`.
 */
export const memory = {
  /**
   * Store text in memory.
   * Returns the generated entry ID.
   */
  store(text: string, category: string): string {
    if (isWasmEnv()) {
      return hostBindings()["prx:host/memory"].store(text, category);
    }
    // stub: return a fake ID
    return `stub-${Date.now()}`;
  },

  /**
   * Recall memories matching a query.
   * Returns up to `limit` entries.
   */
  recall(query: string, limit: number): MemoryEntry[] {
    if (isWasmEnv()) {
      return hostBindings()["prx:host/memory"].recall(query, limit);
    }
    return [];
  },
} as const;
